### PR TITLE
Testing Different Clustering Algorithms

### PR DESCRIPTION
**Pull request by marilynzhang**
Beyond DBSCAN, we also tried Agglomerative Clustering, OPTICS, and HDBSCAN as possible clustering algorithms. We were able to increase our recall from 95% to 96% by using HDBSCAN with optimized parameters.

Support for different clustering algorithms:

- Wrapper classes for sklearn's clustering algorithms so that we can create a dependency injection with an already-initialized clustering object passed into find icon, while allowing us to set the default hyperparameters for these clustering objects to what is optimal for our use cases. 
- Update optimizer to reflect the change in structure

### CODE REVIEW

File: modules/algorithms.py [hunk 1]
```diff
@@ -69,45 +69,35 @@ def detect_contours(
...
        The contours and masks are np.ndarrays of List[List[int]]
        and List[bool].
   """
-  clusters = sklearn.cluster.DBSCAN(
-      eps=eps, min_samples=min_samples).fit(image_contours)
-  # a label of -1 means the point was not clustered by DBSCAN - a "noise" point
+  clusters = clusterer.fit(image_contours)
+  # a label of -1 means the point was not clustered - a "noise" point
   n_clusters = len(set(
       clusters.labels_)) - (1 if -1 in clusters.labels_ else 0)
```
Comment (ewadkins) [inline modules/algorithms.py:92]: Let's do something like n_clusters = len(label for label in set(clusters.labels_) if label != -1). Makes it more clear what the logic is doing, and is checking for -1 in the deduplicated list of labels, rather than the list of all labels.

Also, with the above change this shouldn't be necessary, but just a tip for the future. Try to line break at a higher level if possible. E.g.:

Eh:
```
n_clusters = len(set(
      clusters.labels_)) - (1 if -1 in clusters.labels_ else 0)
```

A little better (adding unnecessary parentheses is okay if it makes it more readable):
```
n_clusters = (len(set(clusters.labels_))
              - (1 if -1 in clusters.labels_ else 0))
```

File: modules/algorithms.py [hunk 2]
```diff
@@ -69,45 +69,35 @@ def detect_contours(
...
-      eps: The maximum distance a point can be away to be considered
-       within neighborhood of another point. (default: {10})
-      min_samples: The number of points needed within a neighborhood
-       of a point for it to be a core point. (default: {5})
 
   Returns:
       Tuple: (List of groups of points each representing its own contour,
        List of corresponding boolean mask of core points).
        The contours and masks are np.ndarrays of List[List[int]]
        and List[bool].
```
Comment (ewadkins) [inline modules/algorithms.py:87]: Update return value?

File: modules/clustering_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  def __init__(self,
+               eps: float = 7.5,
+               min_samples: int = 2,
+               metric: str = "euclidean",
+               metric_params: Any = None,
+               algorithm: str = "auto",
+               leaf_size: int = 30,
+               p: float = None,
+               n_jobs: int = None):
```
Comment (ewadkins) [inline modules/clustering_algorithms.py:39]: I know these classes have a lot of params, but that's all the more reason to document them. It'll make our lives much easier down the line :) so please do.

Same for the below classes.

File: modules/icon_finder_shape_context.py [hunk 4]
```diff
@@ -14,22 +15,20 @@ class IconFinderShapeContext(modules.icon_finder.IconFinder):  # pytype: disable
   """This class generates bounding boxes via Shape Context Descriptors."""
 
   def __init__(self,
+               clusterer: clustering_algorithms.
+               SklearnClusterer = clustering_algorithms.DBSCANClusterer(),
```
Comment (ewadkins) [inline modules/icon_finder_shape_context.py:19]: Can you indent this 4 spaces?

File: modules/clustering_algorithms.py (hunk 3 above)
Comment (swongly) [reply modules/clustering_algorithms.py:39]: Since the init functions are just wrappers around some algorithm, and the set of default parameters, you can define a couple of abstract classmethods

One that returns default algorithm parameters,
One that returns the desired algorithm constructor.

Then the init in the abstract base class can be something like...
def __init__(self, *args, **kwargs):
  self._method()(self.default_args or **kwargs)

Comment (ewadkins) [reply modules/clustering_algorithms.py:39]: What might be better actually is just linking to the sklearn function's documentation, and then only adding explicit args for the defaults you're overriding (rather than using the library defaults, e.g. eps). Then, catch all the other args with **kwargs.

File: modules/icon_finder_shape_context.py (hunk 4 above)
Comment (marilynzhang) [reply modules/icon_finder_shape_context.py:19]: (Also discussed over chat) - did not do so due to linter error after indenting 4 spaces.

**Review by ewadkins — APPROVED** [review]

**Review by swongly — APPROVED** [review]
